File: Functions/TypeScript/MainMethod.ts
// Only for Bubble Sort and its variations
export function swap(arr: number[], xp: number, yp: number) {
  var temp = arr[xp];
  arr[xp] = arr[yp];
  arr[yp] = temp;
}

class PrintArray {
  public static normalArray(array: number[]) {
    console.log("Before Sorting: " + "\n" + array + "\n");
  };

  public static sortedArray(array: number[]) {
    console.log("After Sorting: " + "\n" + array + "\n");
  };
}

class RandomArray {
  public static randomArray(size: number): number[] {
    let array: number[] = [];
    for (let i = 0; i < size; i++) {
      array.push(Math.floor(Math.random() * 25600));
    }
    return array;
  };
}

class TimeTaken {
  public static timeTaken(callback: Function): string {
    const startTime = Date.now();
    callback();
    const endTime = Date.now();
    return `Time taken: ${endTime - startTime} milliseconds`;
  }
}

export function mainMethod(callback: Function) {
  let array = RandomArray.randomArray(4096);
  PrintArray.normalArray(array);
  const time = TimeTaken.timeTaken(() => callback(array));
  PrintArray.sortedArray(array);
  console.log(time);
}

export function mainMethodTwo(callback: Function) {
  let array = RandomArray.randomArray(4096);
  PrintArray.normalArray(array);
  const time = TimeTaken.timeTaken(() => callback(array, array.length));
  PrintArray.sortedArray(array);
  console.log(time);
}

File: Functions/Java/MainMethod.java
package Java;

import java.util.Arrays;

public class MainMethod {
  public static void normalArray(int array[]) {
    System.out.println("Before Sorting: " + "\n" + Arrays.toString(array) + "\n");
  }

  public static void sortedArray(int array[]) {
    System.out.println("After Sorting: " + "\n" + Arrays.toString(array) + "\n");
  }

  public static int[] randomArray() {
    int size = 4096 * 4;
    int[] array = new int[size];
    for (int i = 0; i < size; i++) {
      array[i] = (int) (Math.random() * 25600);
    }
    return array;
  }

  public static void timeTaken(long startTime, long endTime) {
    System.out.println("Time taken: " + (endTime - startTime) + " milliseconds");
  }

  // Only for BubbleSort and its variants
  public static void swap(int[] arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }
}
